src/applications/detection/main.cpp

```diff
@@ -51,236 +52,353 @@ using boost::timer;
   }
   std::ofstream outputFile;
-  outputFile.open( outputFileName );
-  
-  if ( (ext == ".png") || (ext == ".jpg") || (ext == ".PNG") || (ext == ".JPG")) {
-    
+  outputFile.open(outputFileName);
+
+  if((ext == ".png") || (ext == ".jpg"))
+  {
+
     std::cout << "******************* Image mode **********************" << std::endl;
-    
+
     POP_INFO("looking at image " << myPath.string());
-    
-    // Gray scale convertion
+
+    // Gray scale conversion
     cv::Mat src = cv::imread(cmdline._filename);
     cv::Mat graySrc;
-    cv::cvtColor( src, graySrc, CV_BGR2GRAY );
+    cv::cvtColor(src, graySrc, CV_BGR2GRAY);
 
     const int pipeId = 0;
+    boost::ptr_list<CCTag> markers;
 #ifdef PRINT_TO_CERR
-    detection(0, pipeId, graySrc, params, bank, std::cerr, myPath.stem().string());
+    detection(0, pipeId, graySrc, params, bank, markers, std::cerr, myPath.stem().string());
 #else
-    detection(0, pipeId, graySrc, params, bank, outputFile, myPath.stem().string());
+    detection(0, pipeId, graySrc, params, bank, markers, outputFile, myPath.stem().string());
 #endif
-} else if (ext == ".avi" )
+  }
+  else if(ext == ".avi" || ext == ".mov" || useCamera)
   {
     CCTAG_COUT("*** Video mode ***");
-    POP_INFO( "looking at video " << myPath.string() );
+    POP_INFO("looking at video " << myPath.string());
 
     // open video and check
-    cv::VideoCapture video( cmdline._filename.c_str() );
+    cv::VideoCapture video;
+    if(useCamera)
+      video.open(std::atoi(cmdline._filename.c_str()));
+    else
+      video.open(cmdline._filename);
+    
     if(!video.isOpened())
     {
-      CCTAG_COUT("Unable to open the video : " << cmdline._filename); return -1;
+      CCTAG_COUT("Unable to open the video : " << cmdline._filename);
+      return EXIT_FAILURE;
     }
 
-    // play loop
-    int lastFrame = video.get(CV_CAP_PROP_FRAME_COUNT);
-
-    std::list<cv::Mat*> frames;
+    const std::string windowName = "Detection result";
+    cv::namedWindow(windowName, cv::WINDOW_NORMAL);
 
     std::cerr << "Starting to read video frames" << std::endl;
-    while( video.get(CV_CAP_PROP_POS_FRAMES) < lastFrame )
+    std::size_t frameId = 0;
+    
+    // time to wait in milliseconds for keyboard input, used to switch from
+    // live to debug mode
+    int delay = 10;
+
+    while(true)
     {
       cv::Mat frame;
       video >> frame;
-      cv::Mat* imgGray = new cv::Mat;
+      if(frame.empty())
+        break;
       
-      if( frame.channels() == 3 || frame.channels() == 4 )
-        cv::cvtColor(frame, *imgGray, cv::COLOR_BGR2GRAY);
+      cv::Mat imgGray;
+
+      if(frame.channels() == 3 || frame.channels() == 4)
+        cv::cvtColor(frame, imgGray, cv::COLOR_BGR2GRAY);
       else
-        frame.copyTo(*imgGray);
+        frame.copyTo(imgGray);
 
-      frames.push_back( imgGray );
-    }
-    std::cerr << "Done. Now processing." << std::endl;
+      boost::timer t;
```

Comment:
Shouldn't the timer be *inside* the `for`-loop if the meaning is to measure per-frame time? You should also instantiate it just before calling `detection`. Could also upgrade to use the new timer: http://www.boost.org/doc/libs/1_49_0/libs/timer/doc/cpu_timers.html